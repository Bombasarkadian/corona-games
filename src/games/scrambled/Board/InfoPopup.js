import React, { useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";
import { Icon, Popup } from "semantic-ui-react";

import { BONUSES } from "../config";

import styles from "./InfoPopup.module.scss";

const propTypes = {
  popupHandleRef: PropTypes.object,
  open: PropTypes.bool,
  errors: PropTypes.arrayOf(PropTypes.string),
  newWords: PropTypes.arrayOf(PropTypes.object),
  specialBonus: PropTypes.bool,
};

const InfoPopup = ({ popupHandleRef, open, errors, newWords, specialBonus }) => {
  const { t } = useTranslation("scrambled");

  const onMouseMove = useCallback(({ x, y }) => {
    // Viewport relative coordinates
    const popup = document.getElementById("words-info-popup");
    if (popup) {
      const rect = popup.getBoundingClientRect();
      popup.style.transition = "opacity 0.3s, transform 0.3s";
      if (
        x >= rect.left - 10 &&
        x <= rect.right + 10 &&
        y >= rect.top - 10 &&
        y <= rect.bottom + 10
      ) {
        popup.style.opacity = 0;
      } else {
        popup.style.opacity = 1.0;
      }
    }
  }, []);

  useEffect(() => {
    window.addEventListener("mousemove", onMouseMove);
    return () => {
      window.removeEventListener("mousemove", onMouseMove);
    };
  }, [onMouseMove]);

  return (
    <Popup
      context={popupHandleRef}
      position="top center"
      open={open}
      offset={[0, 5]}
      style={{ opacity: 0.95, border: 0, borderRadius: 10, background: "#f5f5f5" }}
      popper={{
        id: "words-info-popup",
        style: { pointerEvents: "none", transition: "opacity 0.3s" },
      }}
      flowing
      basic
      header={errors.length > 0 ? t("game.errors.header") : t("game.errors.ok")}
      content={
        errors.length > 0 ? (
          errors.map((error) => {
            return (
              <p key={error}>
                <Icon color="orange" name="warning sign" /> {t(`game.errors.${error}`)}
              </p>
            );
          })
        ) : (
          <>
            {newWords.map(({ letters, points, wordBonuses }, idx) => {
              return (
                <div key={idx} className={styles.letter_wrapper}>
                  <Icon color="green" name="check circle" />
                  {letters.map((letter, jdx) => (
                    <span key={`${idx}-${jdx}`} className={styles.letter}>
                      {letter}
                    </span>
                  ))}
                  {wordBonuses.map((multiply, jdx) => (
                    <span
                      key={`${idx}-${jdx}`}
                      className={styles.bonus_wrapper}
                      style={{
                        background: BONUSES.word[multiply],
                      }}
                    >
                      x{multiply}
                    </span>
                  ))}
                  <span className={styles.points_wrapper}>{points}</span>
                </div>
              );
            })}
            {specialBonus && (
              <div className={styles.letter_wrapper}>
                <Icon color="yellow" name="winner" />
                {t("game.info.bonus")}
                <span
                  className={styles.bonus_wrapper}
                  style={{
                    background: "#16ab39",
                  }}
                >
                  +50
                </span>
              </div>
            )}
          </>
        )
      }
    />
  );
};

InfoPopup.propTypes = propTypes;
export default InfoPopup;
