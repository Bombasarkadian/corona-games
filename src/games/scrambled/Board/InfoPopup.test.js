import React from "react";
import { render, screen } from "utils/test/render";
import InfoPopup from "./InfoPopup";

describe("scrambled InfoPopup component", () => {
  it("renders correctly", async () => {
    expect.hasAssertions();
    const { container } = render(
      <InfoPopup errors={[]} popupHandleRef={{}} newWords={[]} specialBonus={true} clickable />
    );
    expect(await screen.findByText("game.errors.ok")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders errors correctly", async () => {
    expect.hasAssertions();
    const { container } = render(
      <InfoPopup
        errors={["test_error"]}
        popupHandleRef={{}}
        newWords={[]}
        specialBonus={true}
        clickable
      />
    );
    expect(await screen.findByText("game.errors.header")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
