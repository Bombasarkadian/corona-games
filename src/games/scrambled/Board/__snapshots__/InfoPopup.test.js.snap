// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrambled InfoPopup component renders correctly 1`] = `
<div
  id="words-info-popup"
  style="pointer-events: auto; transition: opacity 0.3s; display: flex; position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, 0px);"
>
  <div
    class="ui top center basic flowing popup transition visible"
    style="position: initial; opacity: 0.95; border: 0px; border-radius: 10px; background: rgb(245, 245, 245);"
  >
    <div
      class="header"
    >
      game.errors.ok
    </div>
    <div
      class="letter_wrapper"
    >
      <i
        aria-hidden="true"
        class="yellow winner icon"
      />
      game.info.bonus
      <span
        class="bonus_wrapper"
        style="background: rgb(22, 171, 57);"
      >
        +50
      </span>
    </div>
  </div>
</div>
`;

exports[`scrambled InfoPopup component renders errors correctly 1`] = `
<div
  id="words-info-popup"
  style="pointer-events: auto; transition: opacity 0.3s; display: flex; position: absolute; left: 0px; top: 0px; bottom: 0px; transform: translate(0px, 0px);"
>
  <div
    class="ui top center basic flowing popup transition visible"
    style="position: initial; opacity: 0.95; border: 0px; border-radius: 10px; background: rgb(245, 245, 245);"
  >
    <div
      class="header"
    >
      game.errors.header
    </div>
    <p>
      <i
        aria-hidden="true"
        class="orange warning sign icon"
      />
       
      game.errors.test_error
    </p>
  </div>
</div>
`;
